### ISSUE TITLE
Timeout happening too early

### ISSUE THREAD

**Issue by bergus** [petkaantonov/bluebird#1417]
I'm reporting the case from [this SO question](https://stackoverflow.com/questions/44814204/how-fast-efficient-is-bluebirds-timeout):

> The following example times out in most cases (outputs `timed out`):
> 
>     Promise = require('bluebird');
>     
>     new Promise(resolve => {
>         setTimeout(resolve, 1000);
>     })
>         .timeout(1001)
>         .then(() => {
>             console.log('finished');
>         })
>         .catch(error => {
>             if (error instanceof Promise.TimeoutError) {
>                 console.log('timed out');
>             } else {
>                 console.log('other error');
>             }
>         });
> 
> Does this mean that the Bluebird's promise overhead takes longer than 1ms? I see it often time out even if I use `.timeout(1002)`.
> 
> ---
> 
> Using Bluebird 3.5.0, under Node.js 8.1.2

I would expect this to *never* output `timed out`, since the timeouts happen in order, and the latter with higher timespan value. Given that the promise chain resolution happens in the microtask queue, the fulfillment callback should log `finished` before the macrotask timer callback from `.timeout()` runs and even get a chance at cancelling the promise.

Is this a bug, or are my expectations off?

**Comment by Ginden** [comment]
> Is this a bug, or are my expectations off?

That seems like race condition. Look at this code: https://gist.github.com/a31229454897fcbebc1aed6be9e38325

And output: 
```js
in setTimeout, p is resolved, timingP is pending
nextTick, p is resolved, timingP is pending
.isPending called by bluebird/src/timers.js#L75, p is resolved, timingP is pending
after timeout, p is resolved, timingP is rejected
```

* calling `resolve` does not resolve promise synchronously - not sure if it's expected behaviour
* in nextTick `p` is resolved, but this didn't propagate to `timingP`
* function planted by `setTimeout` runs. It checks if `timingP` is resolved - but it isn't, so `timingP` become rejected

I will probably create pull request for this.

@petkaantonov How should I write test for non-deterministic behaviour like that?

**Comment by petkaantonov** [comment]
Yes you are expecting too much setTimeout. There is never any real guarantee that they are called separate at different times. E.g. If you `setTimeout(..., 1000)` and `setTimeout(..., 100000)` and then hide the tab in mobile and return 10 seconds later, both callbacks should be called at the same time (synchronously, with no async in-between). The fix is it not to rely on setTimeout like that.

**Comment by sanji-programmer** [comment]
I can confirm that the fix 892286b can avoid the race reported by @bergus; I could reproduce the race with a tool I've worked with to test different schedules of the event loop.

Nevertheless, if we do the opposite:
```
const Promise = require('bluebird');

new Promise(resolve => {
    setTimeout(function timeout() { resolve() }, 1001);
})
    .timeout(1000)
    .then(function then() {
        console.log('finished');
    })
    .catch(function catch_(error) {
        if (error instanceof Promise.TimeoutError) {
            console.log('timed out');
        } else {
            console.log('other error');
        }
    });
```
the race remains with this code giving 'time out' or 'finished'  based on the status of event loop. 

It seems that, at least in Node, callback `timeout()` in this code and callback `timeoutTimeout()` in timers.js are registered in the same tick of event loop; so some order between them can be defined. The issue may be that `this.drainQueues()` (line 16 of async.js) is registered using `setImmediate()` and it can race with the other timeout callbacks.

**Comment by benjamingr** [comment]
@sanji-programmer Node timers are not accurate to the millisecond fwiw. There is an open bug for that I can hunt it if it's important to you.

**Comment by sanji-programmer** [comment]
@benjamingr tks for the info; if it is not too much work, I would like to see this open bug. 

As fix 892286b seemed to solve the former case, I feel like to report, maybe this case is also fixable.
